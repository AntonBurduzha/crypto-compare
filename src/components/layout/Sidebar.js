// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Layout, Menu, Icon } from 'antd';
import * as navActions from '../../actions';
import type { Action } from '../../types/actions';
import type { StoreState } from '../../types/reducers';
import type { Element } from 'react';

type StateProps = { tab: string };

type DispatchProps = { getNextTab: (string) => void };

type State = { collapsed: boolean };

type Props = StateProps & DispatchProps;

type NavListObject = { key: string, icon: string, text: string };

export class Sidebar extends React.Component<Props, State> {
	navList: Array<NavListObject> = [
		{ key: 'list', icon: 'desktop', text: 'Crypto List' },
		{ key: 'marketBTC', icon: 'area-chart', text: 'BTC Market' },
		{ key: 'marketETH', icon: 'area-chart', text: 'ETH Market' }
	];
	state = { collapsed: false };

	onCollapse = (collapsed: boolean): void => this.setState({ collapsed });

	onChangeTab = (type: string): Function => (): void => this.props.getNextTab(type);

	render() {
		return (
			<Layout.Sider collapsible collapsed={this.state.collapsed} onCollapse={this.onCollapse}>
				<Menu theme="dark" defaultSelectedKeys={[this.props.tab]} mode="inline">
					{ this.navList.map((nav: NavListObject): Element<any> => {
						return (
							<Menu.Item key={nav.key} onClick={this.onChangeTab(nav.key)}>
								<Icon type={nav.icon}/><span>{nav.text}</span>
							</Menu.Item>
						);
					})
				}
				</Menu>
			</Layout.Sider>
		);
	}
}

function mapStateToProps(state: StoreState): StateProps {
	return { tab: state.app.navigation.tab }
}

function mapDispatchToProps(dispatch: Action => void): DispatchProps {
	return bindActionCreators({ ...navActions }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(Sidebar);
