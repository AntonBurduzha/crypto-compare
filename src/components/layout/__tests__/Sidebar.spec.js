import React from 'react';
import { shallow, mount } from 'enzyme';
import { Layout, Menu, Icon } from 'antd';
import { Sidebar, mapStateToProps, mapDispatchToProps }  from '../Sidebar';

describe('Sidebar component', () => {
  const spy = jest.fn();
  const props = { getNextTab: spy }
  const wrap = shallow(<Sidebar {...props}/>);

  it('renders without crashing', () => {
    expect(wrap).toBeDefined();
  });

  it('has 3 tabs', () => {
    expect(wrap.find(Menu).children()).toHaveLength(3);
  });

  it('has icons in each tab', () => {
    expect(wrap.find(Menu.Item).find(Icon).exists()).toBeTruthy();
    expect(wrap.find(Menu.Item).find(Icon).length).toEqual(3);
  });

  it('checks state and it\'s changing', () => {
    wrap.setState({ collapsed: false });
    expect(wrap.state()).toEqual({ collapsed: false });

    wrap.find(Layout.Sider).props().onCollapse(true);
    expect(wrap.state()).toEqual({ collapsed: true });
  });

  it('emit action when user click on tab', () => {
    wrap.find(Menu.Item).at(1).simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith('marketBTC');
  });

  it('has redux context', () => {
    const initState = {
      app: { navigation: { tab: 'list' } }
    };
    const dispatch = jest.fn();

    expect(mapStateToProps(initState).tab).toEqual('list');
    expect(mapDispatchToProps(dispatch).getNextTab).toBeDefined();
  });

  it('matches snapshot', () => {
    expect(wrap).toMatchSnapshot();
  })
})
