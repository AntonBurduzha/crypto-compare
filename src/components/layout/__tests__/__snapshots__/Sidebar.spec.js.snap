// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Sidebar
    getNextTab={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Menu
        className=""
        defaultSelectedKeys={
                Array [
                        undefined,
                      ]
        }
        focusable={false}
        mode="inline"
        prefixCls="ant-menu"
        theme="dark"
>
        <MenuItem
                onClick={[Function]}
        >
                <Icon
                        type="desktop"
                />
                <span>
                        Crypto List
                </span>
        </MenuItem>
        <MenuItem
                onClick={[Function]}
        >
                <Icon
                        type="area-chart"
                />
                <span>
                        BTC Market
                </span>
        </MenuItem>
        <MenuItem
                onClick={[Function]}
        >
                <Icon
                        type="area-chart"
                />
                <span>
                        ETH Market
                </span>
        </MenuItem>
</Menu>,
      "collapsed": true,
      "collapsedWidth": 80,
      "collapsible": true,
      "defaultCollapsed": false,
      "onCollapse": [Function],
      "prefixCls": "ant-layout-sider",
      "reverseArrow": false,
      "style": Object {},
      "theme": "dark",
      "width": 200,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <MenuItem
            onClick={[Function]}
>
            <Icon
                        type="desktop"
            />
            <span>
                        Crypto List
            </span>
</MenuItem>,
          <MenuItem
            onClick={[Function]}
>
            <Icon
                        type="area-chart"
            />
            <span>
                        BTC Market
            </span>
</MenuItem>,
          <MenuItem
            onClick={[Function]}
>
            <Icon
                        type="area-chart"
            />
            <span>
                        ETH Market
            </span>
</MenuItem>,
        ],
        "className": "",
        "defaultSelectedKeys": Array [
          undefined,
        ],
        "focusable": false,
        "mode": "inline",
        "prefixCls": "ant-menu",
        "theme": "dark",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "list",
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Icon
                type="desktop"
/>,
              <span>
                Crypto List
</span>,
            ],
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "type": "desktop",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Crypto List",
              },
              "ref": null,
              "rendered": "Crypto List",
              "type": "span",
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "marketBTC",
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Icon
                type="area-chart"
/>,
              <span>
                BTC Market
</span>,
            ],
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "type": "area-chart",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "BTC Market",
              },
              "ref": null,
              "rendered": "BTC Market",
              "type": "span",
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "marketETH",
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Icon
                type="area-chart"
/>,
              <span>
                ETH Market
</span>,
            ],
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "type": "area-chart",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "ETH Market",
              },
              "ref": null,
              "rendered": "ETH Market",
              "type": "span",
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Menu
          className=""
          defaultSelectedKeys={
                    Array [
                              undefined,
                            ]
          }
          focusable={false}
          mode="inline"
          prefixCls="ant-menu"
          theme="dark"
>
          <MenuItem
                    onClick={[Function]}
          >
                    <Icon
                              type="desktop"
                    />
                    <span>
                              Crypto List
                    </span>
          </MenuItem>
          <MenuItem
                    onClick={[Function]}
          >
                    <Icon
                              type="area-chart"
                    />
                    <span>
                              BTC Market
                    </span>
          </MenuItem>
          <MenuItem
                    onClick={[Function]}
          >
                    <Icon
                              type="area-chart"
                    />
                    <span>
                              ETH Market
                    </span>
          </MenuItem>
</Menu>,
        "collapsed": true,
        "collapsedWidth": 80,
        "collapsible": true,
        "defaultCollapsed": false,
        "onCollapse": [Function],
        "prefixCls": "ant-layout-sider",
        "reverseArrow": false,
        "style": Object {},
        "theme": "dark",
        "width": 200,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <MenuItem
              onClick={[Function]}
>
              <Icon
                            type="desktop"
              />
              <span>
                            Crypto List
              </span>
</MenuItem>,
            <MenuItem
              onClick={[Function]}
>
              <Icon
                            type="area-chart"
              />
              <span>
                            BTC Market
              </span>
</MenuItem>,
            <MenuItem
              onClick={[Function]}
>
              <Icon
                            type="area-chart"
              />
              <span>
                            ETH Market
              </span>
</MenuItem>,
          ],
          "className": "",
          "defaultSelectedKeys": Array [
            undefined,
          ],
          "focusable": false,
          "mode": "inline",
          "prefixCls": "ant-menu",
          "theme": "dark",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "list",
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Icon
                  type="desktop"
/>,
                <span>
                  Crypto List
</span>,
              ],
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "type": "desktop",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Crypto List",
                },
                "ref": null,
                "rendered": "Crypto List",
                "type": "span",
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "marketBTC",
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Icon
                  type="area-chart"
/>,
                <span>
                  BTC Market
</span>,
              ],
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "type": "area-chart",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "BTC Market",
                },
                "ref": null,
                "rendered": "BTC Market",
                "type": "span",
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "marketETH",
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Icon
                  type="area-chart"
/>,
                <span>
                  ETH Market
</span>,
              ],
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "type": "area-chart",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "ETH Market",
                },
                "ref": null,
                "rendered": "ETH Market",
                "type": "span",
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
